### ISSUE TITLE
Query regarding MS-COCO Captioning evaluation - CIDEr score reproduction guidance

### ISSUE THREAD

**Issue by undefined443** [M-E-AGI-Lab/Muddit#8]
Dear Authors,

Thank you for your excellent work on Muddit and for making the model publicly available. I've been working on reproducing the captioning benchmark results from your paper. I've completed a partial evaluation on the MS-COCO 2014 dataset and would be very grateful for your guidance on a few details to ensure my setup aligns with yours.

I evaluated Muddit on MS-COCO 2014 validation set captioning using the Muddit-512 checkpoint from Hugging Face. Here is a brief summary of my process:

Environment:

- Model: [MeissonFlow/Muddit](https://huggingface.co/MeissonFlow/Muddit/tree/main/512/transformer) (512×512 checkpoint)
- Dataset: [MS-COCO 2014 validation set](http://images.cocodataset.org/zips/val2014.zip) (40504 images)
- Generation: 32 steps, CFG=1.5, prompt="Please describe this image."

Observations:

The generated captions are fluent and grammatically correct, but the style differs significantly from COCO ground truth annotations, resulting in very low CIDEr scores. I suspect this might be due to an inappropriate prompt on my end.

Questions:

1. Could you please share the evaluation code used to achieve the CIDEr score of 59.9 reported in Table 2 of your paper?
2. If sharing the code is not possible, could you clarify the prompt template or format you used for evaluation?
3. Besides the prompt, are there other critical configuration parameters that affect captioning quality?

Any advice would be incredibly valuable for my work and for others in the community building upon your research.

Thank you again for your time and for this wonderful contribution to the field.

**Comment by Shi-qingyu** [comment]
Thank you for your interest in our work. When testing CIDEr, since COCO captions are relatively short, we used the prompt "Please describe this image concisely." This encourages Muddit to output briefer descriptions, avoiding the issue you mentioned where overly long descriptions result in lower scores. We used `cfg=1.0` and `steps=32`.
The testing script will be provided to you after code review. We recommend simply adding "concisely" to resolve the issue of descriptions being too long.

**Comment by Shi-qingyu** [comment]
Hi! Apologies for the delay. Here is the evaluation script:

```
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
coco_caption_val2014_eval.py
----------------------------

Evaluate image captioning performance on MS-COCO val2014 using UnifiedPipeline.
This script generates captions and computes the CIDEr score incrementally.

Example
-------
python coco_caption_val2014_eval.py \
       --ann      /data/coco2014/annotations/captions_val2014.json \
       --img-dir  /data/coco2014/val2014 \
       --save-pred ./predictions.json \
       --bs 8 --device cuda
"""

import argparse
import json
import math
import logging
import sys
from collections import Counter, defaultdict
from pathlib import Path
from typing import List, Dict, Tuple, Any, Optional

import numpy as np
import torch
from tqdm import tqdm

# ========== 0. Dependencies (UnifiedPipeline) ==========
try:
    from src.pipeline import UnifiedPipeline
    from src.transformer import SymmetricTransformer2DModel
    from src.scheduler import Scheduler
    from train.trainer_utils import load_images_to_tensor
    from transformers import CLIPTextModelWithProjection, CLIPTokenizer
    from diffusers import VQModel
except ImportError as e:
    print(f"[ERROR] Failed to import project dependencies: {e}")
    print("Please ensure you are running this script from the project root.")
    sys.exit(1)

# Configure Logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(message)s",
    datefmt="%H:%M:%S"
)
logger = logging.getLogger(__name__)


# ========== 1. Pipeline Builder ==========

def build_pipe(model_path: str, ckpt_path: Optional[str], device: str) -> UnifiedPipeline:
    """
    Initialize and load the UnifiedPipeline model.
    """
    logger.info(f"Loading model from {model_path}...")
    
    tokenizer = CLIPTokenizer.from_pretrained(model_path, subfolder="tokenizer")
    text_encoder = CLIPTextModelWithProjection.from_pretrained(model_path, subfolder="text_encoder")
    vqvae = VQModel.from_pretrained(model_path, subfolder="vqvae")
    
    # Load transformer from checkpoint if provided, otherwise from base model
    transformer_path = ckpt_path if ckpt_path else model_path
    logger.info(f"Loading transformer from {transformer_path}")
    transformer = SymmetricTransformer2DModel.from_pretrained(
        transformer_path, subfolder="transformer"
    )
    
    scheduler = Scheduler.from_pretrained(model_path, subfolder="scheduler")

    pipe = UnifiedPipeline(
        vqvae=vqvae,
        tokenizer=tokenizer,
        text_encoder=text_encoder,
        transformer=transformer,
        scheduler=scheduler
    )
    return pipe.to(device)


# ========== 2. CIDEr Scorer ==========

class CiderScorer:
    """
    Handles CIDEr score computation using either pycocoevalcap (preferred)
    or a NumPy fallback implementation.
    """
    def __init__(self):
        self.use_official = False
        try:
            from pycocoevalcap.cider.cider import Cider
            self.scorer = Cider()
            self.use_official = True
            logger.info("[INFO] Using official pycocoevalcap for CIDEr.")
        except ImportError:
            logger.warning("[WARN] pycocoevalcap not found. Using NumPy fallback implementation.")
            self.use_official = False

    def compute_score(self, refs: Dict[int, List[str]], hyps: Dict[int, str]) -> float:
        """
        Compute CIDEr score.
        
        Args:
            refs: Dictionary mapping image_id to list of reference captions.
            hyps: Dictionary mapping image_id to a single generated caption.
        """
        # Align keys
        common_ids = refs.keys() & hyps.keys()
        if not common_ids:
            logger.warning("No common image IDs found between references and hypotheses.")
            return 0.0

        # Filter dictionaries
        refs_filtered = {i: refs[i] for i in common_ids}
        hyps_filtered = {i: [hyps[i]] for i in common_ids} # pycocoevalcap expects list of strings

        if self.use_official:
            # Suppress internal logging from pycocoevalcap if needed, or keep it
            score, _ = self.scorer.compute_score(refs_filtered, hyps_filtered)
            return score
        else:
            return self._compute_cider_numpy(refs_filtered, hyps_filtered)

    # --- NumPy Fallback Implementation ---
    def _precook(self, s: str, n: int = 4) -> Counter:
        cnt = Counter()
        words = s.split()
        for k in range(1, n + 1):
            for i in range(len(words) - k + 1):
                cnt[tuple(words[i:i + k])] += 1
        return cnt

    def _compute_idf(self, refs: List[List[str]], n: int = 4) -> Dict[tuple, float]:
        df = defaultdict(int)
        for ref_list in refs:
            # Flatten n-grams for this image's references
            ngrams = set()
            for r in ref_list:
                counts = self._precook(r, n)
                ngrams.update(counts.keys())
            for ng in ngrams:
                df[ng] += 1
        
        N = len(refs)
        return {ng: math.log(max(1.0, N) / dfv) for ng, dfv in df.items()}

    def _cider_score_single(self, hyp: str, refs: List[str], idf: Dict, n: int = 4) -> float:
        hyp_cnt = self._precook(hyp, n)
        ref_cnts = [self._precook(r, n) for r in refs]
        
        score = 0.0
        for k in range(1, n + 1):
            # Vectorize hypothesis
            hk = np.array([hyp_cnt.get(ng, 0) for ng in idf if len(ng) == k])
            if hk.sum() > 0:
                hk = hk / np.linalg.norm(hk)
            
            # Vectorize references and compute similarity
            sim_k = []
            for ref_cnt in ref_cnts:
                rk = np.array([ref_cnt.get(ng, 0) for ng in idf if len(ng) == k])
                if rk.sum() > 0:
                    rk = rk / np.linalg.norm(rk)
                sim_k.append(np.minimum(hk, rk).sum()) # Clipping logic
            
            score += np.mean(sim_k) if sim_k else 0.0
            
        return score * 10.0  # Standard scaling

    def _compute_cider_numpy(self, refs: Dict[int, List[str]], hyps: Dict[int, List[str]]) -> float:
        # 1. Compute IDF over all references
        all_refs_list = list(refs.values())
        idf = self._compute_idf(all_refs_list)
        
        # 2. Compute score per image
        scores = []
        for img_id in refs:
            hyp_str = hyps[img_id][0] # Assuming single hypothesis
            ref_list = refs[img_id]
            scores.append(self._cider_score_single(hyp_str, ref_list, idf))
            
        return float(np.mean(scores))


# ========== 3. Main Logic ==========

def generate_batch(batch: List[Tuple[int, Path]], pipe: UnifiedPipeline, cfg: argparse.Namespace, hyps: Dict[int, str]):
    """
    Run inference on a batch of images.
    """
    img_ids = [item[0] for item in batch]
    img_paths = [item[1] for item in batch]

    try:
        # Load images
        imgs = load_images_to_tensor(img_paths, target_size=(cfg.res, cfg.res)).to(cfg.device)
        
        # Inference
        prompts = ["Please describe this image concisely."] * len(batch)
        
        outs = pipe(
            prompt=prompts,
            image=imgs,
            height=cfg.res,
            width=cfg.res,
            guidance_scale=cfg.cfg,
            num_inference_steps=cfg.steps
        )

        # Store results
        for img_id, caption in zip(img_ids, outs.prompts):
            if img_id not in hyps:
                hyps[img_id] = caption

    except Exception as e:
        logger.error(f"Error processing batch {img_ids}: {e}")


@torch.no_grad()
def main(cfg: argparse.Namespace):
    # Set seed for reproducibility
    if cfg.seed is not None:
        torch.manual_seed(cfg.seed)
        np.random.seed(cfg.seed)

    # Load Annotations
    logger.info(f"Loading annotations from {cfg.ann}...")
    with open(cfg.ann, 'r') as f:
        coco = json.load(f)

    # Map image_id -> file_name
    id2file = {img["id"]: img["file_name"] for img in coco["images"]}
    img_ids = sorted(id2file.keys())
    logger.info(f"Total images in dataset: {len(img_ids):,}")

    # Build Pipeline
    pipe = build_pipe(cfg.model, cfg.ckpt, cfg.device)

    # Prepare References
    refs = defaultdict(list)
    for ann in coco["annotations"]:
        refs[ann["image_id"]].append(ann["caption"])

    # Initialize Scorer (moved before loop for intermediate scoring)
    scorer = CiderScorer()

    # Generation Loop
    hyps = {}
    batch = []
    batch_cnt = 0
    
    # Debug limit
    target_ids = img_ids[:cfg.limit] if cfg.limit > 0 else img_ids

    logger.info(f"Starting generation for {len(target_ids)} images...")
    
    for img_id in tqdm(target_ids, desc="Generating Captions"):
        img_path = Path(cfg.img_dir) / id2file[img_id]
        
        if not img_path.exists():
            logger.warning(f"Image not found: {img_path}, skipping.")
            continue

        batch.append((img_id, img_path))

        if len(batch) >= cfg.bs:
            generate_batch(batch, pipe, cfg, hyps)
            batch = [] # Reset batch
            batch_cnt += 1

            # === Intermediate Scoring every 20 batches ===
            if batch_cnt % 20 == 0:
                current_score = scorer.compute_score(refs, hyps)
                tqdm.write(f"[Batch {batch_cnt}] Current CIDEr: {current_score * 100:.2f}")

    # Process remaining items
    if batch:
        generate_batch(batch, pipe, cfg, hyps)

    # Save Predictions
    if cfg.save_pred:
        output_data = [{"image_id": int(i), "caption": c} for i, c in hyps.items()]
        with open(cfg.save_pred, "w") as f:
            json.dump(output_data, f, indent=2)
        logger.info(f"[✓] Predictions saved to {cfg.save_pred}")

    # Final CIDEr Score
    final_score = scorer.compute_score(refs, hyps)
    
    print("\n" + "="*30)
    print(f"Final CIDEr Score: {final_score * 100:.2f}")
    print("="*30 + "\n")


# ========== 4. CLI Entry Point ==========

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="COCO Validation Evaluation Script")
    
    # Data arguments
    parser.add_argument("--ann",      required=True, help="Path to captions_val2014.json")
    parser.add_argument("--img-dir",  required=True, help="Directory containing COCO val2014 images")
    parser.add_argument("--save-pred", default=None, help="Path to save generated predictions (json)")
    
    # Model arguments
    parser.add_argument("--model",    default="MeissonFlow/Meissonic", help="Base model path or HuggingFace ID")
    parser.add_argument("--ckpt",     default=None, help="Path to specific transformer checkpoint")
    parser.add_argument("--device",   default="cuda", help="Device to use (cuda/cpu)")
    
    # Inference arguments
    parser.add_argument("--bs",       type=int, default=32, help="Batch size")
    parser.add_argument("--res",      type=int, default=512, help="Image resolution")
    parser.add_argument("--steps",    type=int, default=32, help="Number of inference steps")
    parser.add_argument("--cfg",      type=float, default=1.0, help="Classifier Free Guidance scale")
    parser.add_argument("--seed",     type=int, default=42, help="Random seed for reproducibility")
    
    # Debugging
    parser.add_argument("--limit",    type=int, default=-1, help="Limit number of images for debugging (-1 for all)")

    args = parser.parse_args()
    
    main(args)
```

**Comment by undefined443** [comment]
Thank you so much for the detailed explanation and the evaluation script! 

The tip about using "Please describe this image concisely." to encourage shorter descriptions is very helpful.

Really appreciate your time and effort!
